File: iosApp/iosApp/RootScreen.swift
import SwiftUI
import Shared
import KMMViewModelCore
import KMMViewModelSwiftUI

struct RootScreen: View {
    @StateViewModel var viewModel = RootViewModel()
    
    var body: some View {
        VStack {
            switch viewModel.state {
            case is RootState.ApplicationNotActivated: AppNotActivatedScreen()
            case is RootState.ApplicationNotLoggedIn:
                LoginScreen()
            case is RootState.ApplicationLoggedIn:
                KMMDemoScreen()
            case is RootState.Loading:
                Text("Root Loading")
            default:
                Text("Loading")
            }
        }
    }
}

struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: composeApp/src/commonMain/kotlin/root/RootScreen.kt
package root

import LandingViewModel
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import demo.KMMDemoScreen
import landing.AppNotActivatedScreen
import login.LoginScreen
import org.koin.androidx.compose.getViewModel
import vm.login.LoginViewModel
import vm.root.RootState
import vm.root.RootViewModel

@Composable
fun RootScreen(
    viewModel: RootViewModel = getViewModel()
) {
    val state by viewModel.state.collectAsState()
    when (state) {
        is RootState.ApplicationLoggedIn -> {
            KMMDemoScreen()
        }

        is RootState.ApplicationNotLoggedIn -> {
            val loginViewModel: LoginViewModel = getViewModel()
            val loginUiState by loginViewModel.state.collectAsState()
            LoginScreen(
                loginUiState,
                onPinUpdate = { loginViewModel.onPinUpdate(it) },
                onOkClicked = { loginViewModel.onOkButtonClicked() })
        }

        RootState.ApplicationNotActivated -> {
            val landingViewModel: LandingViewModel = getViewModel()
            AppNotActivatedScreen(onActivateClick = { landingViewModel.activateApp() })
        }

        RootState.Loading -> {
            Text("Loading TODO")
        }
    }
}
